sasmodels/kernel_iq.c

```diff
@@ -129,34 +168,52 @@ static void set_spin_weights(double in_spin, double out_spin, double weight[6])
 static double mag_sld(
   const unsigned int xs, // 0=dd, 1=du.real, 2=ud.real, 3=uu, 4=du.imag, 5=ud.imag
   const double qx, const double qy,
-  const double px, const double py,
+  const double px, const double py,const double pz,
   const double sld,
   const double mx, const double my, const double mz
 )
 {
+  double qvector[3];
+  double Mvector[3];
+  double Pvector[3];
+  double Mperp[3];
+  double MperpP[3];
+  double MperpPperpQ[3];
+  const double qsq = sqrt(qx*qx + qy*qy) ;
+
+  SET_VEC(qvector, qx/qsq, qy/qsq, 0);	
+	
+  SET_VEC(Mvector, mx, my, mz);
+  SET_VEC(Pvector, px, py, pz);
+  ORTH_VEC(Mperp, Mvector, qvector);
+  ORTH_VEC(MperpP, Mperp, Pvector);
+  ORTH_VEC(MperpPperpQ, MperpP, qvector);
```

Comment:
This is called six times for each pixel on the detector; it might run a little faster if you only compute what you need in the different conditions.

Testing on my machine the difference from the old version is not significant, so maybe it's fine:
```sh
    $ sascomp core_shell_sphere -magnetic -neval=10 -nq=400 -single!   # CPU
    $ sascomp core_shell_sphere -magnetic -neval=10 -nq=400 -single   # GPU
```